Update to website "Create a FF app" tutorial (#7296)
microsoft/FluidFramework: docs/content/docs/start/tutorial.md

  }
```
### Loading an existing dice flow

The second path in the application is much more straight forward. The default data is already set, the container is already attached, and the `id` is passed into the function, rather than returned.

```js

const loadExistingDice = async (id) => {
  const { container } = await client.getContainer(id, containerConfig);
  renderDiceRoller(root, container.initialObjects.diceMap);
}

```

### Switching between loading and creating

Since the application can be in one of two states, creating, or loading from an `id`, the app simulate those states by storing the container `id` in the URL hash. If the URL has a hash, the app will load that existing container, otherwise the app creates a new container and sets the returned `id` as the hash.

Both load and create methods are async, so the `start` function needs to be created and then called, catching any errors that are returned.

```js

async function start() {
  if (location.hash) {
    await loadExistingDice(location.hash.substring[1])
  } else {
    const id = await createNewDice();
    location.hash = id;
  }
}
start().catch((error) => console.error(error));

```

## Defining the Dice view

> Fluid is framework agnostic and works well with React, Vue, Angular and web components. This example will use nothing more than standard HTML/DOM methods, but visit the [HelloWorld repo](https://github.com/microsoft/FluidHelloWorld/tree/main/src/view) to see this view written in other frameworks. 

Reviewer: ```suggestion
Fluid is view framework agnostic and works well with React, Vue, Angular and web components. This example will use standard HTML/DOM methods, but visit the [HelloWorld repo](https://github.com/microsoft/FluidHelloWorld/tree/main/src/view) to see this view written in other frameworks. 
```

### Start with a static view

> In this tutorial we will start with a dice roller that works without Fluid locally, and then show how to add collaboration by changing a few key pieces. 

Reviewer: ```suggestion
It is simplest to create the view using local data without Fluid, then add Fluid by changing some key pieces of the app. This tutorial uses this approach.
```

> This `renderDiceRoller`, given an HTML element to attach to, creates a working dice roller that displays a random dice value each time the "Roll" button is clicked. Note that demo code ommits the styles for brevity.

Reviewer: ```suggestion
This `renderDiceRoller` function, given an HTML element to attach to, creates a working dice roller that displays a random dice value each time the "Roll" button is clicked. Note that the included code snippets omit the styles for brevity.
```

```js
function renderDiceRoller(elem, diceMap) {
    const dice = document.createElement("div");

    const rollButton = document.createElement("button");
    rollButton.textContent = "Roll";
    rollButton.onclick = () => updateDice(Math.floor(Math.random() * 6));
    
    elem.append(dice, rollButton);

    const updateDice = (value) => {
        // Unicode 0x2680-0x2685 are the sides of a die (⚀⚁⚂⚃⚄⚅).
        dice.textContent = String.fromCodePoint(0x2680 + value);
    };
    updateDice(1);
}
```

### Modifying Fluid data

> The first thing to change is what happens when the user clicks the `rollButton`. Instead of having the button update the state directly, the button updates the `value` key on the passed in `diceMap`. This creates a new op on the client which will be communicated with all of the other collaborators. Pushing local state out to Fluid state is a common pattern because the view should react to local and remote changes in the same way.

Reviewer: ```suggestion
To begin using Fluid in the application, the first thing to change is what happens when the user clicks the `rollButton`. Instead updating the local state directly, the button updates the `value` key on the passed in `diceMap`. This change will be distributed to all clients and cause a `valueChanged` event to be sent. The event handler will then update the view. This pattern is common in Fluid because the view should react to local and remote changes in the same way.
```

```js
    rollButton.onclick = () => diceMap.set("value", Math.floor(Math.random() * 6));
```

### Relying on Fluid data

> Secondly, the `updateDice` function will no longer take in an arbitrary value. The app can no longer directly modify the local dice value. 

Reviewer: ```suggestion
The next change that needs to be made is to change the `updateDice` function so it no longer accepts an arbitrary value. This means the app can no longer directly modify the local dice value. Instead, the value will be retrieved from the `SharedMap` each time `updateDice` is called.
```

```js
    const updateDice = () => {
        const value = diceMap.get("value");
        dice.textContent = String.fromCodePoint(0x2680 + value);
    };
    updateDice();
```

> ### Listening for Fluid changes

Reviewer: ```suggestion
### Handling remote changes
```

> The value returned from `diceMap.get("value")` is only a snapshot in time. To keep the data up to date as it changes an event listener is set on the `diceMap` to call `updateDice` each time that the `valueChanged` event is fired. Visit each DDSes docs to see a list of events and the values passed to those events. 

Reviewer: ```suggestion
The value returned from `diceMap.get("value")` is only a snapshot in time. To keep the data up to date as it changes an event handler is set on the `diceMap` to call `updateDice` each time that the `valueChanged` event is sent. See the documentation for individual DDSes to see a list of events and the values passed to those events. 
```

  - "/start/tutorial/"
---

In this walkthrough, you'll learn about using the Fluid Framework by building a simple [DiceRoller](https://github.com/microsoft/FluidHelloWorld) application. To get started, go through the [Quick Start]({{< relref "quick-start.md" >}}) guide.

{{< fluid_bundle_loader idPrefix="dice-roller"
    bundleName="dice-roller.12142020.js" >}}

In the DiceRoller app, users are shown a die with a button to roll it. When the die is rolled, Fluid Framework syncs the data across clients so everyone sees the same result. To do this, complete the following steps:

> 1. Application setup

Reviewer 2: To reflect Tyler's change below.

```suggestion
1. Set up the application
```

2. Create a Fluid Container
3. Write the dice view
4. Connect view to Fluid data

## Application setup

To create a Fluid application that connects to the [Azure Fluid Relay service]({{< relref "azure-frs.md" >}}), start by creating a new instance of the Azure Fluid Relay service client. The client is responsible for creating and loading containers as well as communicating changes between all of the collaborators. Learn more about configuring the client at [Azure Fluid Relay service client]({{< relref "azure-frs.md" >}}).

The app creates Fluid containers using a configuration that defines a set of *initial objects* that will be available in the container. Learn more about initial objects in [Data modeling]({{< relref "data-modeling.md" >}}).

Lastly, `root` defines the HTML element that the Dice will render on.

```js
import { AzureClient } from "@fluidframework/azure-client";
import { SharedMap } from "fluid-framework";

> const client = new AzureClient(localConfig);

Reviewer 2: `localConfig` doesn't seem to be defined in this context

Author: the previous demo omitted some of the boilerplate code required to set up the actual application. I link to the Azure FRS docs for those needing to set this up on their own.

      initialObjects: { diceMap: SharedMap }
  };

const root = document.getElementById("root")
```

## Create a Fluid container

Fluid data is stored within containers, and these containers need to be created before other users can load them. Since creation and loading of containers both happen in the browser, a Fluid application needs to be capable of handling both paths.

### Create a new container

The create path of the application starts with calling `createContainer` and passing in a schema defining which shared objects will be available on the new `container`.

> After setting some default values on the `diceMap`, the container is attached by calling the `attach` function. Attaching allows changes to be sent to and from the client, and returns the assigned container id that will be used in the load path.

Author: "client" so far has been used for tinylicious client, and "service" hasn't really been used at all. Other "collaborators"? attached to a "client"?

> After setting some default values on the `diceMap`, the container is attached by calling the `attach` function. Attaching allows changes to be sent to and from the client, and returns the assigned container id that will be used in the load path.

Reviewer 3: "Service" is most correct here.  "Attach" is kind of a weird verb for what it means but it's what we have right now.

> After setting some default values on the `diceMap`, the container is attached by calling the `attach` function. Attaching allows changes to be sent to and from the client, and returns the assigned container id that will be used in the load path.

Author: should we instead be importing `tinylicious-service`? Calling it a service but interfacing it through a `const client` make it pretty confusing.

> After setting some default values on the `diceMap`, the container is attached by calling the `attach` function. Attaching allows changes to be sent to and from the client, and returns the assigned container id that will be used in the load path.

Reviewer 3: The service is remote, running on a server in the cloud, so it wouldn't make sense to import the service itself.  Instead you are importing a (local) client object which knows how to communicate with that service.

> After setting some default values on the `diceMap`, the container is attached by calling the `attach` function. Attaching allows changes to be sent to and from the client, and returns the assigned container id that will be used in the load path.

Reviewer 2: Fluid requires a service. To connect to a specific service you need to use the corresponding *-client package. Attaching is the act of creating/registering a new container with the corresponding service instance. 

@tylerbutler, your comment feels a bit backwards because until you attach there's not way to get the id to have another editor. Maybe something like...

> Changes to the dice value will be local until the container is *attached*. Once attached, the app begins sending and receiving operations to and from the service. The `attach` call returns the `id` of the container which other clients can use to load and simultaneously edit against the container.

> Now that the app has a connected container and default data, the dice roller view is ready to render.

Reviewer: ```suggestion
The app now has an attached container that contains data. The final step to complete the app is to create a view to display the dice to the user.
```

```js
function renderDiceRoller(diceMap, elem) {
    const dice = document.createElement("div");

    const rollButton = document.createElement("button");
    rollButton.textContent = "Roll";
    rollButton.onclick = () => updateDice(Math.floor(Math.random() * 6));
    
    elem.append(dice, rollButton);

    const updateDice = (value) => {
        // Unicode 0x2680-0x2685 are the sides of a die (⚀⚁⚂⚃⚄⚅).
        dice.textContent = String.fromCodePoint(0x2680 + value);
    };
    updateDice(1);
}
```

## Connect the view to Fluid data

### Modifying Fluid data

To begin using Fluid in the application, the first thing to change is what happens when the user clicks the `rollButton`. Instead of updating the local state directly, the button updates the number stored in the `value` key of the passed in `diceMap`. Because the `diceMap` is a Fluid `SharedMap`, changes will be distributed to all clients. Any changes to the `diceMap` will cause a `valueChanged` event to be emitted, and an event handler can trigger an update of the view.

> Pushing local state out to Fluid shared objects is a common pattern that allows the view to react the same way for both local and remote changes.

Reviewer: ```suggestion
This pattern is common in Fluid because it enables the view to behave the same way for both local and remote changes.
```

  - "/start/tutorial/"
---

> In this walkthrough, you'll learn about using the Fluid Framework by building a simple [DiceRoller](https://github.com/microsoft/FluidHelloWorld) application. To get started, go through the [Quick Start]({{< relref "quick-start.md" >}}) guide.

Reviewer 2: I feel like in general the doc can use more guidance relative to the actual repo. Some Examples:

- After you run the Quick Start launch the root folder with an IDE.
- In this following section we will be looking at `./src/app.ts`
- In the following section we will be looking at `./src/view/jsView.ts`

Author: good point. I haven't really connected it to the repo code at all, partly because it hasn't been updated yet :) But I'll make some changes to fix that.

Reviewer 2: ```suggestion
In this walkthrough, you'll learn about using the Fluid Framework by examining the DiceRoller application at [https://github.com/microsoft/FluidHelloWorld](https://github.com/microsoft/FluidHelloWorld). To get started, go through the [Quick Start]({{< relref "quick-start.md" >}}) guide.
```

{{< fluid_bundle_loader idPrefix="dice-roller"
    bundleName="dice-roller.12142020.js" >}}

> In the DiceRoller app, users are shown a die with a button to roll it. When the die is rolled, Fluid Framework syncs the data across clients so everyone sees the same result. To do this, complete the following steps:

Reviewer: I've been using *the Fluid Framework*, and that's what I prefer. But I care a lot more about consistency and @Rick-Kirkham 's input. :)

> In the DiceRoller app, users are shown a die with a button to roll it. When the die is rolled, Fluid Framework syncs the data across clients so everyone sees the same result. To do this, complete the following steps:

Reviewer 4: I prefer adding the "The".

Now that the app has a connected container and default data, the dice roller view is ready to render.

```js
const createNewDice = async () => {
    const { container } = await client.createContainer(containerConfig);
    // Set default data
    container.initialObjects.diceMap.set("value", 1);
    // Attach container to service and return assigned ID
    id = container.attach();
    // load the dice roller
    renderDiceRoller(root, container.initialObjects.diceMap);
    return id;
  }
```
### Loading an existing container

Loading a container is more straightforward than creating a new one. When loading, the container already contains data, and is already attached, so those steps are irrelevant. You need only to pass the `id` of the container you wish to load in the `getContainer()` function.

```js
const loadExistingDice = async (id) => {
  const { container } = await client.getContainer(id, containerConfig);
  renderDiceRoller(root, container.initialObjects.diceMap);
}

```

### Switching between loading and creating

> The application supports both creating a new container and loading an existing container using its `id`. To control which state the app is in, it stores the container `id` in the URL hash. If the URL has a hash, the app will load that existing container, otherwise the app creates a new container, attaches it, and sets the returned `id` as the hash.

Reviewer: If you don't mark `id` as code, then it should be capitalized **ID**.

```js
function renderDiceRoller(elem, diceMap) {
    const dice = document.createElement("div");

    const rollButton = document.createElement("button");
    rollButton.textContent = "Roll";
    rollButton.onclick = () => updateDice(Math.floor(Math.random() * 6));
    
    elem.append(dice, rollButton);

    const updateDice = (value) => {
        // Unicode 0x2680-0x2685 are the sides of a die (⚀⚁⚂⚃⚄⚅).
        dice.textContent = String.fromCodePoint(0x2680 + value);
    };
    updateDice(1);
}
```

## Connect the view to Fluid data

### Modifying Fluid data

To begin using Fluid in the application, the first thing to change is what happens when the user clicks the `rollButton`. Instead updating the local state directly, the button updates the number stored in the `value` key of the passed in `diceMap`. This change will be distributed to all clients and cause a `valueChanged` event to be sent. The event handler is then used to trigger an update of the view.

> Pushing local state out to Fluid data is a common pattern because the view should react to local and remote changes in the same way.

Reviewer 2: ```suggestion
Pushing local state out to Fluid shared objects is a common pattern that allows the view to react the same way for both local and remote changes.
```

```js
    rollButton.onclick = () => diceMap.set("value", Math.floor(Math.random() * 6));
```

### Relying on Fluid data

The next change that needs to be made is to change the `updateDice` function so it no longer accepts an arbitrary value. This means the app can no longer directly modify the local dice value. Instead, the value will be retrieved from the `SharedMap` each time `updateDice` is called.

```js
    const updateDice = () => {
        const value = diceMap.get("value");
        dice.textContent = String.fromCodePoint(0x2680 + value);
    };
    updateDice();
```

### Handling remote changes

> The values returned from `diceMap` are only a snapshot in time. To keep the data up to date as it changes an event handler must be set on the `diceMap` to call `updateDice` each time that the `valueChanged` event is sent. See the documentation for individual DDSes to get a list of events fired and the values passed to those events.

Reviewer 2: @tylerbutler , what's the syntax to link to SharedMap docs here?

```suggestion
The values returned from `diceMap` are only a snapshot in time. To keep the data up to date as it changes an event handler must be set on the `diceMap` to call `updateDice` each time that the `valueChanged` event is sent. See the documentation for [`SharedMap`]() to get a list of events fired and the values passed to those events.
```
